### ISSUE TITLE
Type of sample of `Exponential{T}` is not `T`

### ISSUE THREAD

**Issue by wouterwln** [JuliaStats/Distributions.jl#1902]
MWE:
```julia
θ = 1.0f0
d = Exponential(θ)
typeof(rand(Exponential(θ))) == typeof(θ)
```
returns `false`

Because Exponential sampling is also used when sampling `Gamma` with shape 1, the problem also occurs there.

**Comment by bvdmitri** [comment]
This is problematic here
```
julia> Gamma(1f0, 1f0) |> rand
1.1458127367196822

julia> Gamma(2f0, 1f0) |> rand
1.662515f0
```

**Comment by andreasnoack** [comment]
Please search the issue tracker for several similar issues. I think this issue might be redundant.

**Comment by andreasnoack** [comment]
Which version of Distributions are you using? On master I'm getting
```julia
julia> typeof(rand(Exponential(θ))) == typeof(θ)
true
```
and that is a consequence of the commit that you link to (which I had forgotten).

**Comment by wouterwln** [comment]
You are right, on master this seems to be fixed. Reading the discussion in https://github.com/JuliaStats/Distributions.jl/pull/1885 improved my understanding. The issue that is found by our tests is that `eltype(Gamma(1.0f0, 1.0f0))` is not `Float32` (for reference, for `Normal` this returns `Float32`). The story [here](https://github.com/JuliaStats/Distributions.jl/blob/b219803a0d03a7c75d7aef7c0bab6cd0d79997dc/docs/src/extends.md?plain=1#L7) implies that `eltype` should return the type of the samples. Is that correct? If this is not the case, the problem is downstream and we have to make an exception downstream.

**Comment by bvdmitri** [comment]
Indeed for `Gamma` eltype fallbacks to 
```julia
Base.eltype(::Type{<:Sampleable{F,Continuous}}) where {F} = Float64
```
and for `Normal` eltype is defined as
```julia
Base.eltype(::Type{Normal{T}}) where {T} = T
```
in our code base we basically implicitly tested that `eltype(distribution) == typeof(rand(distribution))` and it failed. I would suggest to add those tests in Distributions.jl as well because now the implementation for `Gamma` does not follow the documented behaviour of `eltype(::Type{Sampleable})`.  Specifically this line _This is the type of elements of the
  samples generated by the rand method._ which is not for `Gamma` (and maybe other distributions)
```julia
julia> eltype(Gamma(1f0, 1f0)) == typeof(rand(Gamma(1f0, 1f0)))
false
```
Also this behaviour leads to this inconsistency, where semantically similar code returns different container types
```
julia> rand(Gamma(1f0, 1f0), 1)
1-element Vector{Float64}:
 1.2745195627212524

julia> [ rand(Gamma(1f0, 1f0)) ]
1-element Vector{Float32}:
 3.234097
```

**Comment by andreasnoack** [comment]
> in our code base we basically implicitly tested that `eltype(distribution) == typeof(rand(distribution))`

I think it's best to avoid `eltype(::Distribution)` completely. Please refer to the many long discussion in the issue tracker. As an example of the issue here consider the case

```julia
julia> eltype(Normal(1//1, 1//1))
Rational{Int64}

julia> rand(Normal(1//1, 1//1))
0.3227598879915683
```

so `eltype(::Normal)` is just lying to you.

**Comment by bvdmitri** [comment]
You cannot expect users to refer to long discussion in the issue tracker. The issue tracker is for developers, not for users. Users are looking at the documentation of the package. And the documentation is misleading. If the documentation is lying it must be changed.

**Comment by bvdmitri** [comment]
As a side note, obtaining the exact sample type without having to call `eltype(rand(...))` is indeed a highly useful feature for preallocating containers, as the documentation suggests. However, it seems like this functionality is currently somewhat broken. For example, even with your code:

```julia
julia> distribution = Normal(1//1, 1//1)
Normal{Rational{Int64}}(μ=1//1, σ=1//1)

julia> container = zeros(eltype(distribution), 10);

julia> Distributions.rand!(distribution, container)
ERROR: MethodError: no method matching randn(::Random.TaskLocalRNG, ::Type{Rational{Int64}})
```

This issue makes it challenging to write generic code that preallocates containers for samples from arbitrary distributions, as `eltype` doesn’t seem to be designed for this purpose and is "lying". Is there an alternative method to handle this?

**Comment by andreasnoack** [comment]
My point is that this is an open issue and part of the discussion here is just repeating arguments from other issues.

> obtaining the exact sample type without having to call `eltype(rand(...))` is indeed a highly useful feature for preallocating containers

This is the eternal challenge. There is no easy way to know this type. You can

1. Compute the elements and promote on container as you go
2. Ask Julia's type inference (please don't do this)
3. Try to guess

When you try to write generic code then 3 almost always end up being wrong at some point. Part of the issue here is that Distributions was original written just for Float64 and Int64 parameters and random variates and then later made generic which invalidated some of the original assumptions. But now I'm repeating explanations already available in other issues.

**Comment by bvdmitri** [comment]
I understand your point, but the documentation is a primary source of information for most users (and not GitHub issues) and seems to be incorrect in its current form. If the previous original assumptions were invalidated it should be reflected in the documentation too. Now the documentation states that `eltype` returns the type of the samples, which isn't accurate. Would it be possible to at the very least just update the documentation? This will certainly prevent confusion and the need for repeated explanations in other issues.
